### ISSUE TITLE
Building a Multipart payload in JAX-RS

### ISSUE THREAD

**Issue by andymc12** [OpenLiberty/open-liberty#16031]
## Description of the high level feature, including any external spec links:  
<br/><br/><br/>  
We currently document that it is possible to send and receive multipart/form-data payloads in JAX-RS in Liberty (see https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_jaxrs_multipart_formdata_from_html.html), however, we only currently support receiving multipart/form-data payloads. This is because we only provide the API for reading these payloads - we do not have an API for writing them.

This feature intends to add a builder API that would enable users to create a multipart/form-data payload that could be sent (i.e. sending multiple files to a remote server or sending multiple files back to a client).

The builder API would look something like this:
```
/*******************************************************************************
 * Copyright (c) 2021 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package com.ibm.websphere.jaxrs20.multipart;

public class AttachmentBuilder {
    
    public static AttachmentBuilder newBuilder(String fieldName) {...}

    public AttachmentBuilder contentId(String contentId) {...}

    public AttachmentBuilder contentType(MediaType contentType) {...}

    public AttachmentBuilder contentType(String contentType) {...}

    public AttachmentBuilder header(String headerName, String...headerValues) {...}

    public AttachmentBuilder headers(MultivaluedMap<String, String> newHeaders) {...}

    public AttachmentBuilder fileName(String fileName) {...}

    public AttachmentBuilder inputStream(InputStream inputStream) {...}

    // convenience method that effectively returns fileName(fileName).inputStream(inputStream);
    public AttachmentBuilder inputStream(String fileName, InputStream inputStream) {...}

    public IAttachment build() {...}
}
```

The `IAttachment` interface already exists and is the API used for reading `multipart/form-data` payloads today.

This feature has been requested by multiple customers - most recently at #15896 on behalf of [CQ Framework](https://github.com/cqframework/cql-translation-service/) (discussed in [Slack](https://ibm-cloud.slack.com/archives/C31G324KW/p1613062579040500) - link may not work for users outside IBM).

Note also, that I intend to introduce this pattern into the JAX-RS API as part of the proposed `multipart/form-data` APIs in Jakarta REST 3.1 (or maybe 3.2 if I don't get it in soon enough...)

##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/lyvn0qctxl59hjab9qe0n3ids2095c4u
- FTS: https://github.com/OpenLiberty/open-liberty/issues/16248
- Beta Blog Post (if applicable): Issue https://github.com/OpenLiberty/open-liberty/issues/16249
- Blog Post: 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by covener** [comment]
UFO review notes/comments, most minor detail to be added to content:

 - p9/p10: mention mprestclient impact on as-is/to-be (incl potential API change) [emily]
 - p13: show mime data on the wire [eric]
 - p13: "jaxrs20" quirk and which liberty features would get the support (ee7 and later) [cindy]
 - p13: why differ from jaxrs spec proposal? [emily]
    - IAttachment already used on the receive side
 - p13: inputStream comment on wrong line
 - p13: jaxrs closes InputStream but not until it goes over the wire
 - p13: twas should use same interface if there's a solution there vs. using CXF APIs. [cindy,alasdair]
    - different feature for delivery but not necessaril a separate design [alasdair]
    - sync with p28 for users who found a CXF way in tWAS and moving to Liberty
 - p13: lambda/supplier instead of stream
 - p18: stablization-only [alasdair]
 - p18: feature delivery details for this API and for portable future spec level [emily]
 - p18: application using this API will use spec stuff under the covers in 3.1 and use same 3.1 impl  [cindy]
    - this API shouldn't be removed when we have a 3.1 impl [alasdair]
    - may replace far in the future and require the new interface
 - p22: large file tests for OOM
 - p26: remove queues for public posting of UFO
 - p26: see what it looks like when a stream is closed or throws an exception
 - p28: "custom/IBM javadoc hosting" issue -- will it block delivery? [alasdair]
    - backlog WI exists to fix hosting limitation

**Comment by andymc12** [comment]
@covener thanks for moderating this morning!  I have addressed the feedback in the UFO and uploaded the updated version to https://ibm.box.com/s/lyvn0qctxl59hjab9qe0n3ids2095c4u

**Comment by andymc12** [comment]
This epic should be documented as part of docs issue https://github.com/OpenLiberty/docs/issues/1273.

**Comment by andymc12** [comment]
@samwatibm there are no new messages for this feature so globalization should be **N/A** - would you mind approving this epic for globalization?

@steven1046 likewise, there is no accessibility requirements (**N/A**) for this feature - as it is just an API for application developers. Would you approve for accessibility?

@chirp1 the ID requirements are documented in issue https://github.com/OpenLiberty/docs/issues/1273 - is there anything else you need prior to ID approval?

@jhanders34 any concerns about performance that might block approval?

@tevans78 / @cbridgha I have scheduled a demo for the next EOI meeting. Would you mind approving the Demo requirement?

Thanks all!

**Comment by andymc12** [comment]
UFO: Does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation? **Yes - trace points added to the API methods for EE7, 8 and 9 implementations.**

Test and Demo: As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team). a) What problem paths were tested and demonstrated? b) Who did you demo to? c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? **So far, I only have plans to demo (live) to the EOI meeting next week. STE has been documented and provided to L2 support team. UPDATE: I presented serviceability of four possible issues with this feature to the JAX-RS L3 support/development team - rough video available upon request.**

SVT: SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered. a) Who conducted SVT tests for this feature? b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  **a) Brian Hanczaryk b) Yes**

Service: Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.  **L2: WASWEB (soon to become WASWIN) / L3: WL3 JAX-RS**

Metrics: Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs? **No / N/A**

**Comment by donbourne** [comment]
@andymc12 , purpose of the serviceability demo is to ensure that you go through the problem scenarios with someone so they can help assess whether someone without knowledge of the design/code could get out of problems they may encounter without having to open cases/pmrs.  Demos at EOI generally aren't great for that purpose as they aren't usually focused on assessing this.  So, would be great to see both a demo for that purpose and to get SVT's signoff before I give approval.

**Comment by chirp1** [comment]
David will work on the ID issue and get any further details from Andy as necessary. Approving.

**Comment by donbourne** [comment]
watched video from @andymc12 showing error flows and the related logs.  people in that demo were ok with serviceability when directly asked about it, so taking that as the serviceability demo.
